mail.py

```diff
@@ -0,0 +1,140 @@
+else:
+    regions = ksclient.find_regions()
+
+if not regions:
+    himutils.sys_error('no valid regions found!')
+
+if 'host' in options and options.host:
+    if '.' in options.host:
+        host = options.host
+    else:
+        domain = ksclient.get_config('openstack', 'domain')
+        host = options.host + '.' + domain
+else:
+    host = None
+
+def action_file():
+    if options.template:
+        content = options.template
+        email_content = open(content, 'r')
+        body_content = email_content.read()
+        if options.dry_run:
+            print body_content
+        else:
+            mail = Mail(options.config, debug=options.debug)
+            msg = MIMEText(body_content)
+            msg['subject'] = subject
+            print msg
+            with open(emails_file, 'r') as emails:
+                for toaddr in emails.readlines():
+                    try:
+                        logger.debug('=> Sending email ...')
+                        mail.send_mail(toaddr, msg, fromaddr='noreply@uh-iaas.no')
+                    except ValueError:
+                        himutils.sys_error('Not able to send the email.')
+            emails.close()
+    email_content.close()
+    mail.close()
+
+def action_instance():
+    if options.template:
+        content = options.template
+        email_content = open(content, 'r')
+        body_content = email_content.read()
+        if options.dry_run:
+            print body_content
+        else:
+            with open(content, 'r') as email_content:
+                body_content = email_content.read()
+            for region in regions:
+                novaclient = himutils.get_client(Nova, options, logger, region)
+                instances = novaclient.get_instances()
+                mail = Mail(options.config, debug=options.debug)
+                mail.set_keystone_client(ksclient)
+                users = mail.mail_instance_owner(instances=instances,
+                                                   body=body_content,
+                                                   subject=subject,
+                                                   admin=True)
+    mail.close()
+
+def action_project():
+    if options.template:
+        content = options.template
+        email_content = open(content, 'r')
+        body_content = email_content.read()
+        if options.dry_run:
+            print body_content
+        else:
+            with open(content, 'r') as email_content:
+                body_content = email_content.read()
+            mail = Mail(options.config, debug=options.debug)
+            search_filter = dict()
+            projects = ksclient.get_projects(domain=options.domain, **search_filter)
+            if options.filter and options.filter != 'all':
+                search_filter['type'] = options.filter
+            if options.type:
+                search_filter['type'] = options.type
+            for project in projects:
+                for region in regions:
```

Comment:
Se om du kan bytte om disse for-loopene. Da trenger du kun å initialisere novaclient per region istedet for per region for hvert prosjekt. Du gjør en full auth når du lager novaclienten så jeg vil tro du kan spare minst ett api-kall og det kan bli litt tid om prosjektlisten er lang.